blockstack/lib/nameset/namedb.py

```diff
@@ -1120,7 +1122,57 @@ def get_num_names_in_namespace( self, namespace_id ):
         cur = self.db.cursor()
         return namedb_get_num_names_in_namespace( cur, namespace_id, self.lastblock )
     
+    def get_num_names_in_namespace_at_block( self, namespace_id, block_id ):
+        """
+        Get the number of names in a namespace
+        """
+        cur = self.db.cursor()
+        return namedb_get_num_names_in_namespace( cur, namespace_id, block_id )
     
+    def get_v2_upgrade_threshold_block( self ):
+        """
+        Get the block ID (int) of when the name registration signal threshold was reached.
+        Returns `None` if the threshold has not been reached.
+        """
+        return namedb_get_v2_upgrade_threshold_block(self.db)
+
+    def set_v2_upgrade_threshold_block( self, block_id ):
+        """
+        Set the block ID of when the name registration signal threshold is reached.
+        """
+        namedb_set_v2_upgrade_threshold_block(self.db, block_id)
+
+    def get_v2_import_block_reached( self ):
+        """
+        Get the block ID of when the import block threshold is reached.
+        Returns `None` if the threshold has not been reached.
+        """
+        namedb_get_v2_import_block_reached(self.db)
+
+    def set_v2_import_block_reached( self, block_id ):
+        """
+        Set the block ID of when the name block import threshold is reached.
+        """
+        namedb_set_v2_import_block_reached(self.db, block_id)
+
+    def perform_v2_upgrade_datafile_export( self ):
+        """
+        Export a datafile used for the v2 upgrade.
+        """
+        # TODO: is this called in a way where db locking isn't an issue?
```

Comment:
This function works without any locking issues within the integration environment -- but is there anything to be worried about in prod?

Reply:
Under the hood, it's either (1) copying zone files (which are write-once files), or (2) using a `sqlite3 .backup` subprocess to copy the database (which deals with concurrency issues), so this _should_ be safe.  However, it would be _safest_ if we made this code-path optional (e.g. via a config-file option), so prod wouldn't need to run it.  Instead, we'd have a dedicated VM that ran this code path to produce the export, and we'd NAT it so it isn't serving any public clients.

Reply:
> However, it would be safest if we made this code-path optional (e.g. via a config-file option), so prod wouldn't need to run it.

Done! Env var `V2_MIGRATION_EXPORT` or config file option `v2_migration_export`. Disabled by default.